manpages/imodcmopp.man
.na
.nh
.TH imodcmopp 1 2.00 BL3DFS
.SH NAME
imodcmopp \- color model paint and project a tomogram.
.SH SYNOPSIS
imodcmopp <IMOD contour model> <mrc image> <mrc output file> [options]
.SH DESCRIPTION
imodcmopp makes a color mrc tilt file from different projected
views of tomographic image data contained inside of model
contours.  The model must be in IMOD format and the
tomogram must be in mrc format.
The movie file is in mrc color format. Use the program
frommrc to convert to sgi rgb files.
.SS Options
.TP
.B -x | -z
The 
.B -x 
and 
.B -z 
options are used to change the axis that the 
tilt series will rotate about. The y axis is the default.   
.TP
.B -n
The 
.B -n 
option makes the x and y output size to be large 
enough to fill an NTSC sized window.
.TP
.B -t <first,end,increment>
The 
.B -t 
option is used to change the range and increments in
tilts for the movie.  The default is 
.B -t 0,175,5 
where 0 
degrees is the starting tilt, 175 is the ending tilt and 5 
degrees is the step between the tilts.
.TP
.B -T <tempdir>
imodcmopp creates several temporary files, the defalut directory is
the current working directory, but this option can override this setting.
.SH CAVEATS
imodcmopp needs to have the command xyzproj	in the path in order 
to work.
.SH AUTHOR
Jim Kremer 
.SH SEE ALSO
imodmop, xyzproj, frommrc.
.SH BUGS
The model Z scale information is ignored.
Only available for UNIX machines.
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/03: Modified PIP to read newly invented "autodoc" files to determine
program options and generate manual pages sections.

10/6/03: IMOD_3-0-22
10/6/03: Fixed the default number of views value for transferfid; it was not
being initialized in etomo.

10/6/03: Reordered thread control to eliminate race condition that often 
prevented matchvol1 from running right after any preceeding com script.

10/2/03: IMOD_3-0-21
10/2/03: Modified Rotatevol, Matchvol, and Warpvol to work on volumes up to 
16K in any dimension; Rotatevol was failing for an output X/Y size of 4K by 4K.

10/1/03: Made imodinfo work properly for counting scattered points on subsets
of the volume.

10/1/03: Added message for opening bead fixer window.

10/1/03: Added ability to turn a plugin into an internal module that is added
to the Special menu, to provide Bead fixer on Windows.

9/29/03: Restructured some 3dmod include files to restore ability to build
plugins with minimal header files.

9/25/03: Made sure windows stayed on the screen when restoring Zap and Info
window positions.

9/25/03: Changed Midas to number sections from 1 instead of 0.

9/24/03: IMOD_3-0-20
9/24/03: Resolved problems with the Fortran graphics package so it works
on all systems, at least briefly.

9/24/03: Fixed some problems with event sending under Windows, added a
retry capability to make it work reliably on the Mac, and found that it
did not work with the multi-threaded Qt library on the SGI.

9/24/03: Used different window properties for saving and restoring window
positions, which fixed problems in Redhat 7.3 and Irix.  Made it not restore
Info window size since that is highly font-dependent.

9/23/03: Put Fortran graphics programs inside application directories on the
Mac, which made them work correctly, with the window raisable and resizable.

9/22/03: Removed a large unused common from smoothsurf, which was making it
fail under Cygwin.

9/19/03: Made sample.com files produce top.rec, etc, directly as the output
of Tilt to avoid replacing existing .rec files.

9/19/03: Made 3dmod require only a 15-bit RGB visual, not 16 bit.

9/18/03: IMOD_3-0-19
9/18/03: Added the ability to float (match contrast between sections) based
on a subarea, either the image displayed or the area inside of a rubberband
in the active Zap window.  Also added an autocontrast button to the Info
window with target contrasts settable from the options dialog.

9/17/03: Made the Zap window pan by one display pixel per pixel of mouse move
for zooms less than one.

9/16/03: Made 3dmod save and restore sizes and positions of the Zap and Info
windows, with different settings saved for different image sizes.  Zoom is
adjusted to fit the whole image to the window.  If there are no application
saved settings, the Zap window now opens up wit a size that fits a zoomed-down
image if appropriate.

9/15/03: Eliminated physical flipping of data in 3dmod to avoid very bad
performance when the image data gets into swap space.

9/15/03: Optimized image drawing for zooms equal to the inverse of an integer.

9/13/03: IMOD_3-0-18
9/12/03: Prevented buffer overruns in an array that holds the filename in
3dmod, fixed problems with model checksums that were leading to queries to
save an unchanged model, arranged the startup page in two columns, fixed a
problem with zooming up under Windows, and added a workaround to a display
problem with Qt 3.2.1 on the Mac.

9/11/03: IMOD_3-0-17  beta release with Etomo

9/9/03: IMOD_3-0-16
9/5/03: Added a program, Resamplemod, for rotating a microtubule model to be
perpendicular to the Z axis and creating points at each section or plane
along that axis.

9/4/03: Incorporated changes in Mtlengths to correct for oblique microtubules
in tilted sections and implemented PIP input.

8/29/03: Set up and documented installation procedures on the Mac.

8/29/03: Changed the Fortran graphics routines to run the graphics window in
a separate thread from the Fortran program.  This only works properly under
Linux, where the window refreshes itself when resized or exposed.  On the Mac,
the window can now be moved but not resized.

8/25/03: Added the -E option to Imodmesh to cap the ends of tubes.

8/13/03: IMOD_3-0-15
8/12/03: Made the Qt-based graphics routines used in Fortran programs work on
the Mac without hanging when the window opened.

8/12/03: Made it possible to start up Nda without a model file, just like Mtk.

8/9/03: IMOD_3-0-14
8/9/03: Changed Blendmont to take PIP input and modified command files and

imodutil/imodcmopp.c
/*  IMOD VERSION 2.50
 *
 *  imodcmopp.c --  Create an color mrc image file from a 
 *                  model and an mrc image file.
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 */

/*****************************************************************************
 *   Copyright (C) 1995-2001 by Boulder Laboratory for 3-Dimensional Fine    *
 *   Structure ("BL3DFS") and the Regents of the University of Colorado.     *
 *                                                                           *
 *   BL3DFS reserves the exclusive rights of preparing derivative works,     *
 *   distributing copies for sale, lease or lending and displaying this      *
 *   software and documentation.                                             *
 *   Users may reproduce the software and documentation as long as the       *
 *   copyright notice and other notices are preserved.                       *
 *   Neither the software nor the documentation may be distributed for       *
 *   profit, either in original form or in derivative works.                 *
 *                                                                           *
 *   THIS SOFTWARE AND/OR DOCUMENTATION IS PROVIDED WITH NO WARRANTY,        *
 *   EXPRESS OR IMPLIED, INCLUDING, WITHOUT LIMITATION, WARRANTY OF          *
 *   MERCHANTABILITY AND WARRANTY OF FITNESS FOR A PARTICULAR PURPOSE.       *
 *                                                                           *
 *   This work is supported by NIH biotechnology grant #RR00592,             *
 *   for the Boulder Laboratory for 3-Dimensional Fine Structure.            *
 *   University of Colorado, MCDB Box 347, Boulder, CO 80309                 *
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <mrcfiles.h>
#include "imodel.h"

#define NTSCX 688
#define NTSCY 488

extern struct Mod_Contour *mop_scan(struct Mod_Contour *ocont);
int rgbmodpaint(struct Mod_Model *model, 
                struct MRCheader *hdata, 
                unsigned char **idata, 
                FILE *rtom, FILE *gtom, FILE *btom);
int cmopp_zscale(int zscale, char *fname);
int cmopp_write_mrc(char *moviename, FILE *rfin, FILE *gfin, FILE *bfin);
int cmopp_write_movie(char *moviename, 
                      FILE *rfin, FILE *gfin, FILE *bfin, 
                      char *rgb, int ntsc);
int rgbslice_paint(struct Mod_Model *mod,
                   void *idata,
                   void **pdata,
                   struct MRCheader *hdata,
                   int z);
int contour_paintrgb(struct Mod_Contour *cont,
		     unsigned char *idata,
		     unsigned char **pdata,
		     struct MRCheader *hdata,
		     struct Mod_Object *obj);
int slice_paint(struct Mod_Model *mod,
                void *idata,
                void *pdata,
                struct MRCheader *hdata,
                int z);     
int Mop_red, Mop_green, Mop_blue;
int Cmopp_zscale = 0;
int Cmopp_debug = FALSE;

static void putbyte(FILE *outf, unsigned char val)
{
  unsigned char buf[1];
  buf[0] = val;
  fwrite(buf,1,1,outf);
  return;
}

static void rgb_putshort(FILE *outf, unsigned short val)
{
  unsigned char buf[2];
     
  buf[0] = (val>>8);
  buf[1] = (val>>0);
  fwrite(buf,2,1,outf);
  return;
}

static int putlong(FILE *outf, unsigned int val)
{
  unsigned char buf[4];

  buf[0] = (val>>24);
  buf[1] = (val>>16);
  buf[2] = (val>>8);
  buf[3] = (val>>0);
  return fwrite(buf,4,1,outf);
}

int main( int argc, char *argv[])
{
  FILE *rfin = NULL,  *gfin = NULL,  *bfin = NULL;
  FILE *rfout = NULL, *gfout = NULL,  *bfout = NULL;
  FILE *mfin = NULL;
  FILE *fin = NULL;
  FILE *fout = NULL;
  FILE *of;

  Imod   *imod;
  /*     struct Mod_Model model; */
  struct Mod_Point maxpt;
  struct Mod_Point minpt;
  struct MRCheader hdata;
  struct LoadInfo  li;

  unsigned char **idata;
  unsigned char bdata;
  short sdata;
  short xsize, ysize;
  float ramp, rramp, gramp, bramp;
  int tfill, bfill, lfill, rfill;
  int xmax, ymax, zmax;
  int prosize;
  int i,j,k;
  int pid;
  int mrc = TRUE;
  int xaxis = 0;
  int zaxis = 0;
  int tilt = 0;
  int ntsc = 0;
  int start_tilt = -45;
  int end_tilt  = 45;
  int inc_tilt = 5;

  char command[256];
  char iname[128];
  char tom[64], rtom[64], gtom[64], btom[64],
    rshort[64], gshort[64], bshort[64],
    xyz[64], rxyz[64], gxyz[64], bxyz[64],
    rgb[64];

  char *theTempDir = ".";

  if (argc < 4){
    imodVersion(argv[0]);
    imodCopyright();
    fprintf(stderr,
	    "Usage: %s <model file> <tomogram> <movie file>\n", argv[0]);
    fprintf(stderr, "[-x] [-z] [-n] [-m] [-T <temp>] "
	    "-t <start, end, increment>]\n");
    exit(1);
  }
     
  for (i = 4; i < argc; i++)
    if (argv[i][0] == '-'){
      switch (argv[i][1]){
      case 'd':
	Cmopp_debug = TRUE;
	break;
      case 'n':
	ntsc = TRUE;
	break;
      case 'x':
	xaxis = 1;
	break;
      case 'z':
	zaxis = 1;
	break;
      case 'Z':
	sscanf(argv[++i], "%d", &Cmopp_zscale);
	break;
      case 'm':
	mrc = FALSE;
	break;
      case 't':
	tilt = 1;
	if (argv[i][2] != 0x00)
	  sscanf(argv[i], "-t%d%*c%d%*c%d", &start_tilt,
		 &end_tilt, &inc_tilt);
	else
	  sscanf(argv[++i], "%d%*c%d%*c%d", &start_tilt,
		 &end_tilt, &inc_tilt);
	break;

      case 'T':
	theTempDir = argv[++i];
	break;

      default:
	break;
      }
    }


  pid = getpid();
  sprintf(tom, "%s/%d.tom.mrc",    theTempDir, pid);
  sprintf(rtom, "%s/%d.red.mrc",   theTempDir, pid);
  sprintf(gtom, "%s/%d.green.mrc", theTempDir, pid);
  sprintf(btom, "%s/%d.blue.mrc",  theTempDir, pid);
  sprintf(rshort, "%s/%d.rs.mrc",  theTempDir, pid);
  sprintf(gshort, "%s/%d.gs.mrc",  theTempDir, pid);
  sprintf(bshort, "%s/%d.bs.mrc",  theTempDir, pid);
  sprintf(xyz,    "%s/%d.xyz",     theTempDir, pid);
  sprintf(rxyz,   "%s/%d.rxyz",    theTempDir, pid);
  sprintf(gxyz,   "%s/%d.gxyz",    theTempDir, pid);
  sprintf(bxyz,   "%s/%d.bxyz",    theTempDir, pid);
  sprintf(rgb,    "%s/%d",         theTempDir, pid);

  if ( (xaxis) && (zaxis) ){
    fprintf(stderr, "rgbmop: Only x or z can be selected\n.");
    exit(1);
  }

  if ((start_tilt > end_tilt) && (inc_tilt > 0)){
    fprintf(stderr, "%s: Error, bad tilt input.\n", argv[0]);
    exit(-1);
  }
  if ((end_tilt > start_tilt) && (inc_tilt < 0)){
    fprintf(stderr, "%s: Error, bad tilt input.\n", argv[0]);
    exit(-1);
  }
  if (!inc_tilt){
    fprintf(stderr, "%s: Error, bad tilt input.\n", argv[0]);
    exit(-1);
  }
  if (! (end_tilt - start_tilt)){
    fprintf(stderr, "%s: Error, bad tilt input.\n", argv[0]);
    exit(-1);
  }

  printf("Reading model %s.\n", argv[1]);

  imod = imodRead(argv[1]);
  if (!imod){
    fprintf(stderr, "rgbmop: Couldn't open/read %s\n", argv[1]);
    exit(-1);
  }

  /*     mfin = fopen(argv[1], "r");
	 if (mfin == NULL){
	 fprintf(stderr, "rgbmop: Couldn't open %s\n", argv[1]);
	 exit(-1);
	 }
  */
  gfin = fopen(argv[2], "r");
  if (gfin == NULL){
    fprintf(stderr, "rgbmop: Couldn't open %s\n", argv[2]);
    exit(-1);
  }

  /*     if (ReadImod(&model, mfin)){ */
  /*        fprintf(stderr, "rgbmop: Error reading modelfile %s.\n", argv[1]); */
  /*        exit(-1); */
  /*     } */

  imodel_maxpt(imod, &maxpt);
  imodel_minpt(imod, &minpt);

  if (mrc_head_read(gfin, &hdata)){
    fprintf(stderr, "rgbmop: Can't read tomogram header, %s.\n",
	    argv[2]);
    exit(-1);
  }
     
  li.white = imod->whitelevel;
  li.black = imod->blacklevel;
  li.xmax = hdata.nx;
  li.xmin = 0;
  li.ymax = hdata.ny;
  li.ymin = 0;
  li.zmax = hdata.nz - 1;
  li.zmin = 0;

  idata = mrc_read_byte(gfin, &hdata, &li, mrc_default_status);

  if (!idata){
    fprintf(stderr, "%s: Error reading image data\n", argv[0]);
    exit(-1);
  }
  rfout = fopen(rtom, "w");
  gfout = fopen(gtom, "w");
  bfout = fopen(btom, "w");
     
  rgbmodpaint(imod, &hdata, idata, rfout, gfout, bfout);

  for (k = 0; k < hdata.nz; k++)
    free(idata[k]);
  free(idata);
  fclose(gfin);
     
  fclose(rfout);
  fclose(gfout);
  fclose(bfout);
  fclose(gfin);
  /*     fclose(mfin); */

  fin = fopen(rtom, "r");
  mrc_head_read(fin, &hdata);
  xmax = hdata.nx - 1;
  ymax = hdata.ny - 1;
  zmax = hdata.nz - 1;
  fclose(fin);

  if (xaxis)
    prosize = ymax;
  else if (zaxis){
    if (xmax > ymax)
      prosize = xmax;
    else 
      prosize = ymax;
  }
  else
    prosize = xmax;
     
  prosize /= 4;
  prosize *= 4;
  prosize += 1;

  /***********/
  /* Project */

  if (Cmopp_zscale)
    zmax *= Cmopp_zscale;

  printf("\nProjecting red channel.\n");

  if (Cmopp_zscale)
    cmopp_zscale(Cmopp_zscale, rtom);

  fout = fopen(xyz, "w");
  fprintf(fout, "%s\n%s\n", rtom, rxyz);
  fprintf(fout, "0,%d,0,%d,0,%d\n", xmax, ymax, zmax);
  if (xaxis)
    fprintf(fout, "x\n");
  else if (zaxis)
    fprintf(fout, "z\n");
  else
    fprintf(fout, "y\n"); 
  if (tilt)
    fprintf(fout, "%d,%d,%d\n", start_tilt, end_tilt, inc_tilt);
  else
    fprintf(fout, "0,175,5\n");
  fprintf(fout, "%d\n", prosize);
  fprintf(fout, "1\n0,1\n0\n");
  fclose(fout);
  fout = NULL;
  sprintf(command, "xyzproj < %s > /dev/null", xyz);
  system(command);
  remove(rshort); 

  printf("Projecting green channel.\n");
  if (Cmopp_zscale)
    cmopp_zscale(Cmopp_zscale, gtom);
  fout = fopen(xyz, "w");
  fprintf(fout, "%s\n%s\n", gtom, gxyz);
  fprintf(fout, "0,%d,0,%d,0,%d\n", xmax, ymax, zmax);
  if (xaxis)
    fprintf(fout, "x\n");
  else if (zaxis)
    fprintf(fout, "z\n");
  else     
    fprintf(fout, "y\n");
  if (tilt)
    fprintf(fout, "%d,%d,%d\n", start_tilt, end_tilt, inc_tilt);
  else
    fprintf(fout, "0,175,5\n");
  fprintf(fout, "%d\n", prosize);
  fprintf(fout, "1\n0,1\n0\n");
  fclose(fout);
  fout = NULL; 
  sprintf(command, "xyzproj < %s > /dev/null", xyz);
  system(command);
  remove(gshort); 

  printf("Projecting blue channel.\n");
  if (Cmopp_zscale)
    cmopp_zscale(Cmopp_zscale, btom);

  fout = fopen(xyz, "w");
  fprintf(fout, "%s\n%s\n", btom, bxyz);
  fprintf(fout, "0,%d,0,%d,0,%d\n", xmax, ymax, zmax);
  if (xaxis)
    fprintf(fout, "x\n");
  else if (zaxis)
    fprintf(fout, "z\n");
  else
    fprintf(fout, "y\n");
  if (tilt)
    fprintf(fout, "%d,%d,%d\n", start_tilt, end_tilt, inc_tilt);
  else
    fprintf(fout, "0,175,5\n");
  fprintf(fout, "%d\n", prosize);
  fprintf(fout, "1\n0,1\n0\n");
  fclose(fout);
  sprintf(command, "xyzproj < %s > /dev/null", xyz);
  system(command);
  remove(bshort); 
  remove(xyz);
  remove(rtom);
  remove(gtom);
  remove(btom);


  /*******************************/
  /* write movie file.           */

  printf("creating rgb files\n");     

  rfin = fopen(rxyz, "r");
  gfin = fopen(gxyz, "r");
  bfin = fopen(bxyz, "r");

  if (mrc)
    cmopp_write_mrc(argv[3], rfin, gfin, bfin);
  else
    cmopp_write_movie(argv[3], rfin, gfin, bfin, rgb, ntsc);


  fclose(rfin);
  fclose(gfin);
  fclose(bfin);
  remove(rxyz);
  remove(gxyz);
  remove(bxyz);
  exit(0);
}


int cmopp_write_mrc(char *moviename, FILE *rfin, FILE *gfin, FILE *bfin)
{
  FILE *of;
  struct MRCheader hdata;
  float ramp, rramp = 256.0f, gramp = 256.0f, bramp = 256.0f;
  short sdata;
  unsigned char bdata;
  int i, j, k;
  float min, max, mean;

  of = fopen(moviename, "w");
  if (!of)
    return(-1);

  mrc_head_read(rfin, &hdata);
  if (hdata.amax)
    rramp = 255.0 / hdata.amax;
  mrc_head_read(gfin, &hdata);
  if (hdata.amax)
    gramp = 255.0 / hdata.amax;
  mrc_head_read(bfin, &hdata);
  if (hdata.amax)
    bramp = 255.0 / hdata.amax;
  ramp = rramp;
  if (gramp < ramp)
    ramp = gramp;
  if (bramp < ramp)
    ramp = bramp;

  hdata.mode = MRC_MODE_RGB;
  hdata.swapped = 0;
  mrc_head_write(of, &hdata);
  fseek(rfin, 1024, 0);
  fseek(gfin, 1024, 0);
  fseek(bfin, 1024, 0);
  for (k = 0; k < hdata.nz; k++)
    for(j = 0; j < hdata.ny; j++)
      for(i = 0; i < hdata.nx; i++){
	fread(&sdata, 2, 1, rfin);

	bdata = (unsigned char)((float)sdata * ramp);
	fwrite(&bdata, 1, 1, of);
	fread(&sdata, 2, 1, gfin);

	bdata = (unsigned char)((float)sdata * ramp);
	fwrite(&bdata, 1, 1, of);
	fread(&sdata, 2, 1, bfin);

	bdata = (unsigned char)((float)sdata * ramp);
	fwrite(&bdata, 1, 1, of);
      }

  fclose(of);
  return(0);
}


int cmopp_write_movie(char *moviename, 
                      FILE *rfin, FILE *gfin, FILE *bfin, 
                      char *rgb, int ntsc)
{
  struct MRCheader hdata;
  FILE *of;
  float ramp, rramp = 256.0f, gramp = 256.0f, bramp = 256.0f;
  short xsize, ysize;
  int tfill, bfill, lfill, rfill;
  int i, j, k;
  unsigned char bdata;
  short sdata;
  char command[128],iname[128];

  mrc_head_read(rfin, &hdata);
  if (hdata.amax)
    rramp = 255.0 / hdata.amax;
  mrc_head_read(gfin, &hdata);
  if (hdata.amax)
    gramp = 255.0 / hdata.amax;
  mrc_head_read(bfin, &hdata);
  if (hdata.amax)
    bramp = 255.0 / hdata.amax;
  ramp = rramp;
  if (gramp < ramp)
    ramp = gramp;
  if (bramp < ramp)
    ramp = bramp;
     
  tfill = 0;
  bfill = 0;
  rfill = 0;
  lfill = 0;
  xsize = hdata.nx;
  ysize = hdata.ny;

  if (ntsc){
    if (xsize < NTSCX)
      xsize = NTSCX;
    if (ysize < NTSCY)
      ysize = NTSCY;
  }

  /* keep xsize and ysize multipules of 4 bytes in length. */
  /* outdated, needed for movieplayer. */
  if (xsize % 4)
    xsize += (4 - (xsize % 4));

  if (ysize % 4)
    ysize += (4 - (ysize % 4));

  if (xsize != hdata.nx){
    lfill = (xsize - hdata.nx) / 2;
    rfill = lfill + ((xsize - hdata.nx) % 2 );
  }

  if (ysize != hdata.ny){
    tfill = (ysize - hdata.ny) / 2;
    bfill = tfill + ((ysize - hdata.ny) % 2);
  }

  fseek(rfin, 1024, 0);
  fseek(gfin, 1024, 0);
  fseek(bfin, 1024, 0);
  for (k = 0; k < hdata.nz; k++){
    printf(".");
    fflush(stdout);
    sprintf(iname, "%s.%3.3d.rgb", moviename, k);
          
    if (NULL == (of  = fopen(iname, "w"))){
      fprintf(stderr, "rgbmop: Couldn't open rgb file\n");
      exit(-1);
    }

    rgb_putshort(of, (short)474);       /* MAGIC                */
    putbyte(of, (char)0);          /* STORAGE is VERBATIM  */
    putbyte(of, (char)1);          /* BPC is 1             */
    rgb_putshort(of, (short)3);         /* DIMENSION is 3 */
    rgb_putshort(of, xsize);     /* XSIZE                */
    rgb_putshort(of, ysize);     /* YSIZE                */
    rgb_putshort(of, (short)3);         /* ZSIZE                */
    putlong(of, (long)0);          /* PIXMIN is 0          */
    putlong(of, (long)255);        /* PIXMAX is 255        */
    for(i = 0; i < 4; i++)      /* DUMMY 4 bytes        */
      putbyte(of,0);
    fwrite(iname,80,1,of);  /* IMAGENAME            */
    putlong(of,0l);          /* COLORMAP is 0        */
    for(i = 0; i < 404; i++)    /* DUMMY 404 bytes      */
      putbyte(of,0);

    bdata = 0;
    for (j = 0; j < tfill; j++)
      for (i = 0; i < xsize; i++)
	putbyte(of,0);
    for (j = 0; j < hdata.ny; j++){
      for (i = 0; i < lfill; i++)
	putbyte(of,0);
      for(i = 0; i < hdata.nx; i++){
	fread(&sdata, 2, 1, rfin);
	bdata = (float)sdata * ramp;
	fwrite(&bdata, 1, 1, of);
      }
      bdata = 0;
      for (i = 0; i < rfill; i++)
	putbyte(of,0);
    }
    bdata = 0;
    for (j = 0; j < bfill; j++)
      for (i = 0; i < xsize; i++)
	putbyte(of,0);


    /* Write green data. */
    bdata = 0;
    for (j = 0; j < tfill; j++)
      for (i = 0; i < xsize; i++)
	fwrite(&bdata, 1, 1, of);
    for (j = 0; j < hdata.ny; j++){
      for (i = 0; i < lfill; i++)
	fwrite(&bdata, 1, 1, of);
      for(i = 0; i < hdata.nx; i++){
	fread(&sdata, 2, 1, gfin);
	bdata = (float)sdata * ramp;
	fwrite(&bdata, 1, 1, of);
      }
      bdata = 0;
      for (i = 0; i < rfill; i++)
	fwrite(&bdata, 1, 1, of);
    }
    bdata = 0;
    for (j = 0; j < bfill; j++)
      for (i = 0; i < xsize; i++)
	fwrite(&bdata, 1, 1, of);
          


    /* Write blue data. */
    bdata = 0;
    for (j = 0; j < tfill; j++)
      for (i = 0; i < xsize; i++)
	fwrite(&bdata, 1, 1, of);
    for (j = 0; j < hdata.ny; j++){
      for (i = 0; i < lfill; i++)
	fwrite(&bdata, 1, 1, of);
      for(i = 0; i < hdata.nx; i++){
	fread(&sdata, 2, 1, bfin);
	bdata = (float)sdata * ramp;
	fwrite(&bdata, 1, 1, of);
      }
      bdata = 0;
      for (i = 0; i < rfill; i++)
	fwrite(&bdata, 1, 1, of);
    }
    bdata = 0;
    for (j = 0; j < bfill; j++)
      for (i = 0; i < xsize; i++)
	fwrite(&bdata, 1, 1, of);
          
    fclose(of);
          
  }
  /*
    printf("\ncreating movie file %s\n", moviename);
    sprintf(command, "moviemaker -l 2 -R -o %s %s.???.rgb", moviename, rgb);
    system(command);
    sprintf(command, "/bin/rm -r %s.???.rgb", rgb);
  */
  printf("\n");
  return(0);
}

int rgbmodpaint(struct Mod_Model *model, 
                struct MRCheader *hdata, 
                unsigned char **idata, 
                FILE *rtom, FILE *gtom, FILE *btom)
{
  int xmin, xmax, ymin, ymax, zmin, zmax;
  struct Mod_Point maxpt, minpt;
  struct MRCheader hdout;
  unsigned char **pdata;
  short sdata;
  int i,j,k;
  int xysize;
  imodel_maxpt(model, &maxpt);
  imodel_minpt(model, &minpt);
     
     
  pdata    = (unsigned char **)malloc(3 * sizeof(unsigned char *));
  pdata[0] = (unsigned char *)
    malloc(hdata->nx * hdata->ny * sizeof(unsigned char));
  pdata[1] = (unsigned char *)
    malloc(hdata->nx * hdata->ny * sizeof(unsigned char));
  pdata[2] = (unsigned char *)
    malloc(hdata->nx * hdata->ny * sizeof(unsigned char));

  zmax = hdata->nz - 1;
  zmin = 0;
  ymax = maxpt.y;
  ymin = minpt.y;
  xmax = maxpt.x;
  xmin = minpt.x;

  if ( (minpt.z < hdata->nz) && (minpt.z > 0))
    zmin = minpt.z;
  if ( (maxpt.z < (hdata->nz - 1))  &&  (maxpt.z > minpt.z) )
    zmax = maxpt.z;
     
  mrc_head_new(&hdout, xmax - xmin + 1, ymax - ymin + 1, zmax - zmin + 1,
	       MRC_MODE_SHORT);
  mrc_head_write(rtom, &hdout);
  mrc_head_write(gtom, &hdout);
  mrc_head_write(btom, &hdout);
  xysize = hdata->nx * hdata->ny;

  for( k = zmin; k <= zmax; k++){
    for (i = 0; i < xysize; i++){
      pdata[0][i] = 0;
      pdata[1][i] = 0;
      pdata[2][i] = 0;
    }
    printf("Mopping section %d ", k);
    fflush(stdout);
    rgbslice_paint(model, (void *)&(idata[k][0]), 
		   (void **)pdata, hdata, k); 
    slice_paint(model, NULL, &(pdata[0][0]), hdata, k); 
    slice_paint(model, NULL, &(pdata[1][0]), hdata, k); 
    slice_paint(model, NULL, &(pdata[2][0]), hdata, k); 
          
    for (j = ymin; j <= ymax; j++){
      for (i = xmin; i <= xmax; i++){
	sdata = pdata[0][(j * hdata->nx) + i];
	fwrite(&sdata, 2, 1, rtom);
	sdata = pdata[1][(j * hdata->nx) + i]; 
	fwrite(&sdata, 2, 1, gtom);
	sdata = pdata[2][(j * hdata->nx) + i];
	fwrite(&sdata, 2, 1, btom);
      }
    }
    printf("\r");
  }

  free( (pdata[0]));
  free( (pdata[1]));
  free( (pdata[2]));
  free(pdata);
  return(0);
}


int rgbslice_paint(struct Mod_Model *mod,
                   void *idata,
                   void **pdata,
                   struct MRCheader *hdata,
                   int z)
{

  struct Mod_Object *obj;
  struct Mod_Contour *cont;
  struct Mod_Contour *fcont = NULL;
  int objnum, contnum;

  for (objnum = 0; objnum < mod->objsize; objnum++){
    obj = &(mod->obj[objnum]);
    if (obj->flags & IMOD_OBJFLAG_OPEN)
      continue;
    if (obj->flags & IMOD_OBJFLAG_OUT)
      continue;
    if (obj->flags & IMOD_OBJFLAG_OFF)
      continue;
    for(contnum = 0; contnum < obj->contsize; contnum++){
      cont = &(obj->cont[contnum]);
      if (cont == NULL)
	continue;
      if (cont->pts == NULL)
	continue;
      if (cont->pts[0].z == z){
	fcont = imodel_contour_scan(cont);  
	contour_paintrgb(fcont, (unsigned char *)idata, 
			 (unsigned char **)pdata, hdata, obj);    
      }
      if (fcont) {
	imodel_contour_delete(fcont);  
	fcont = NULL;
      }
    }
  }
  return(0);
}

int contour_paintrgb(struct Mod_Contour *cont,
		     unsigned char *idata,
		     unsigned char **pdata,
		     struct MRCheader *hdata,
		     struct Mod_Object *obj)
{
  int next, lx;
  int i, x, y;
  int xmin, xmax;
  unsigned char mopbyte;
  float r,g,b;
  int bgnpt, endpt, j;

  if (!cont)
    return(-1);
  if (!obj)
    return(-1);
  if (!hdata)
    return(-1);

  for(i = 0; i < cont->psize - 1; i++){
    y = cont->pts[i].y;
    bgnpt = i;
    while (cont->pts[i].y == cont->pts[i+1].y){
      ++i;
      if (i >= cont->psize){
	i = cont->psize - 1;
	break;
      }
    }
    endpt = i;

    /* ERROR: odd number of edge points,      */
    /*        just draw line across all data. */
    if (! ( (endpt-bgnpt)% 2)){
      xmax = cont->pts[bgnpt].x;
      xmin = cont->pts[endpt].x;
      if (xmin >= cont->surf)
	for(x = xmin; 
	    ((x <= xmax)&&( x < hdata->nx)&&(y < hdata->ny)); 
	    x++){
	  mopbyte = idata[x + (y * hdata->nx)];
	  r = (float)mopbyte * obj->red;
	  g = (float)mopbyte * obj->green;
	  b = (float)mopbyte * obj->blue;
	  pdata[0][x + (y * hdata->nx)] = r;
	  pdata[1][x + (y * hdata->nx)] = g;
	  pdata[2][x + (y * hdata->nx)] = b;
                         
	}
    }
    /* even number of edge points. */
    else{
      for(j = bgnpt; j < endpt; j++){
	xmin = cont->pts[j].x;
	xmax = cont->pts[j+1].x;
	if (xmin >= cont->surf)
	  for(x = xmin; 
	      ((x <= xmax)&&( x < hdata->nx)&&(y < hdata->ny));
	      x++){
	    mopbyte = idata[x + (y * hdata->nx)];
	    r = (float)mopbyte * obj->red;
	    g = (float)mopbyte * obj->green;
	    b = (float)mopbyte * obj->blue;
	    pdata[0][x + (y * hdata->nx)] = r;
	    pdata[1][x + (y * hdata->nx)] = g;
	    pdata[2][x + (y * hdata->nx)] = b;
	  }
	j++;
      }    
    }
  }
  return(0);
}

int cmopp_zscale(int zscale, char *fname)
{
  FILE *fp;
  int i,z,k,xysize;
  short sdata;
  unsigned char **idata;
  struct MRCheader hdata;

  fp = fopen(fname, "r+");
  mrc_head_read(fp, &hdata);
  idata = mrc_read_byte(fp, &hdata, NULL, NULL);
  hdata.nz *= zscale;
  xysize = hdata.nz * hdata.ny;
  hdata.swapped = 0;
  mrc_head_write(fp, &hdata);

  for( k = 0; k < hdata.nz; k++){
    for(z = 0; z < zscale; z++){
      for(i = 0; i < xysize; i++){
	/* write short */
	sdata = idata[k][i];
	fwrite(&sdata, sizeof(short), 1, fp);
      }
    }
    free(idata[k]);
  }
  free(idata);
  return(0);
}     
